File: javascript/ch4/currying.js
const R = require('ramda');

function currying() {
  const checkType = R.curry((typeDef, obj) => {
    if (!R.is(typeDef, obj)) {
      const type = typeof obj;
      throw new TypeError(`형식 불일치: [${typeDef}]이어야 하는데, [${type}]입니다.`);
    }
    return obj;
  });

  console.log(checkType(String)('Curry'));
  console.log(checkType(Number)(3));
  // checkType(String)(42);
}

module.exports = {
  currying
};

File: python/python/ch4/currying.py
import pydash

def currying():
    def check_type_for_currying(type_def, obj):
        if not isinstance(obj, type_def):
            raise TypeError(f'형식 불일치: [{type_def}]이어야 하는데, [{type(obj)}]입니다.')
          
        return obj
      
    check_type = pydash.curry(check_type_for_currying)
    
    print(check_type(str)('Curry'))
    print(check_type(int)(3))
    # check_type(str)(42)